File: TopInterview150/Backtracking/Permutations/solution.py
class Solution(object):
    def backtrack(self, result, comb, nums):
        if len(comb) == len(nums):
            result.append(comb[:])# or .append(List(comb))
            return
        
        for i in nums:
            if i in comb:
                continue
            comb.append(i)
            self.backtrack(result, comb, nums)
            comb.pop()

    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        self.backtrack(result, [], nums)
        return result
        

File: TopInterview150/Backtracking/Permutations/solution.cpp
class Solution {
    public:
        vector<vector<int>> permute(vector<int>& nums) {
            vector<vector<int>> result;
            vector<int> comb;
            
            backtrack(result, comb, nums);
            return result;
        }

        void backtrack(
            vector<vector<int>>& result, 
            vector<int>& comb, 
            vector<int> nums
        )
        {
            if (comb.size() == nums.size())
            {
                result.push_back(comb);
                return;
            }

            for (int i : nums)
            {
                if (find(comb.begin(), comb.end(), i) != comb.end()) continue;

                comb.push_back(i);
                backtrack(result, comb, nums);
                comb.pop_back();
            }
        }
    };